### PR TITLE
feat(finetune): support prompt-completion datasets

### LINKED ISSUE #820: [REQ] add prompt-completion dataset support
Parent: #214

## Summary

Add `prompt-completion` as the second supported Unsloth SFT dataset type and introduce explicit runtime dispatch based on `datasets[].type`.

```yaml
datasets:
  - source: organization/question-answer-data
    type: prompt-completion
```

Expected record:

```json
{
  "prompt": "Question: What is a container image?\nAnswer:",
  "completion": " An immutable package containing an application and its dependencies."
}
```

Existing `type: alpaca` configurations must remain fully backward-compatible.

## Motivation

Go currently permits only `alpaca`, while `pkg/finetune/target_unsloth.py` ignores the configured type and always applies the Alpaca formatter. Adding another allowlisted value without Python dispatch would silently process the new schema as Alpaca.

Prompt-completion is the lowest-risk first extension because the frozen Unsloth/TRL stack already preserves `prompt` and `completion`, appends EOS when needed, and creates a completion mask without relying on a model-specific chat template or response-marker detection.

## Proposed changes

- Add a `prompt-completion` dataset type constant.
- Allow `alpaca` and `prompt-completion` in Go validation.
- Refactor the Python path into explicit dataset loading, schema validation, and preparation/dispatch steps.
- Make Python read `datasets[0].type` and reject unknown types rather than falling back to Alpaca.
- Preserve the current Alpaca path exactly:
  - Existing literal prompt.
  - Existing `instruction`, `input`, and `output` contract.
  - Explicit EOS append.
  - Full-sequence loss.
- For `prompt-completion`:
  - Require a non-empty dataset.
  - Require string `prompt` and non-empty string `completion` values.
  - Keep the two columns separate instead of converting them to `text`.
  - Use completion-only loss.
- Perform basic source/schema validation before model and LoRA allocation where possible.
- Add a JSONL fixture, aikitfile example, documentation, and a one-step GPU smoke path.

## Acceptance criteria

- Existing Alpaca aikitfiles require no changes.
- Alpaca rendering remains byte-for-byte equivalent and retains full-sequence loss.
- `type: prompt-completion` is accepted by Go and Python dispatch.
- Prompt-completion records never pass through `format_alpaca_examples`.
- Prompt tokens are masked and completion/EOS tokens are supervised.
- Missing, null, or invalid fields fail with a useful error before expensive model allocation where possible.
- Unknown types fail in both Go and Python.
- A one-step GPU smoke run trains and exports a non-empty GGUF.

## Tests

- Go validation tests for Alpaca, prompt-completion, and unknown types.
- Python tests for runtime dispatch, preserved Alpaca formatting, missing columns, invalid values, and failure ordering.
- Exact-lock preprocessing test proving completion-only labels.
- Add:
  - `test/fixtures/prompt-completion-smoke.jsonl`
  - `test/aikitfile-unsloth-prompt-completion-smoke.yaml`
- Run `make test`, `make test-python`, `make lint`, and the GPU fine-tune/export smoke test.

## Out of scope

- Messages or ShareGPT.
- Raw text.
- Column mappings.
- Alternate splits/subsets/revisions/checksums.
- Multiple datasets.
- Strict YAML behavior changes.
- Response-only chat loss.
- DPO, ORPO, KTO, GRPO, or multimodal objectives.

### PR DESCRIPTION
**Pull request by sozercan**
## Summary

- add explicit `alpaca` and `prompt-completion` dataset dispatch and validation in Go and Python
- preserve byte-for-byte Alpaca formatting while enabling completion-only loss for prompt-completion records
- verify prompt masking, the prompt/completion boundary, packing behavior, and supervised EOS against the active pinned trainer and collator
- add documentation, regression coverage, a JSONL fixture, and a one-step GPU fine-tune/export smoke mode

## Testing

- `make test`
- `PYTHONDONTWRITEBYTECODE=1 make test-python`
- `make check-unsloth-lock`
- `golangci-lint run ./... --timeout 5m --new`
- targeted `actionlint` for `.github/workflows/test-finetune.yaml`
- prompt-completion fixture JSON/YAML validation
- [GPU prompt-completion smoke (attempt 3)](https://github.com/kaito-project/aikit/actions/runs/31113691673): A100 fine-tuning, non-empty GGUF export, CUDA model startup, and chat-completion response validation

Full local `make lint` continues to report the repository's existing `goconst` baseline; changed-lines lint reports zero issues. All PR CI checks pass.

Closes #820

### CODE REVIEW

File: pkg/finetune/target_unsloth.py [hunk 1]
```diff
@@ -186,6 +204,312 @@ def dataset_load_spec(source: str) -> DatasetLoadSpec:
...
+
+    bos_token = getattr(processing_class, "bos_token", None)
+    chat_template = getattr(processing_class, "chat_template", "") or ""
+    add_special_tokens = not (
+        bos_token is not None
+        and (
+            PREPROCESSING_VERIFICATION_PROMPT.startswith(bos_token)
+            or bos_token in chat_template
+        )
+    )
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py]: Addressed against the pinned Unsloth patch. prompt_completion_add_special_tokens mirrors the active Unsloth BOS and chat-template decision, and both source validation and the trainer contract verification use that same setting. The chat-template BOS regression passes, as does GPU smoke run 31156724042 on b7f61b8.

File: pkg/finetune/target_unsloth.py [hunk 2]
```diff
@@ -329,6 +662,9 @@ def train_model(
             output_dir="outputs",
             dataset_text_field="text",
             dataset_num_proc=2,
+            completion_only_loss=(
+                dataset_spec.dataset_type == DATASET_TYPE_PROMPT_COMPLETION
+            ),
             max_length=max_seq_length,
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py:4860]: Addressed in b72b383 and b7f61b8. The source preflight now truncates combined token IDs in the same order as pinned Unsloth, rejects retained sequences with no completion or without a supervised terminal EOS, and the post-preprocessing validator checks every real record or packed segment. Regressions cover both no retained completion and lost EOS. GPU smoke run 31156724042 passed on b7f61b8.

File: pkg/finetune/target_unsloth.py [hunk 3]
```diff
@@ -347,6 +848,26 @@ def train_model(
...
+        validate_prepared_prompt_completion_dataset(
+            trainer.train_dataset,
+            eos_token_id=tokenizer.eos_token_id,
+            max_seq_length=max_seq_length,
+            packing=bool(getattr(trainer.args, "packing", False)),
+            packing_strategy=getattr(trainer.args, "packing_strategy", "bfd"),
+            prompt_prefix_counts=expected_prompt_prefix_counts(
+                dataset,
+                processing_class=tokenizer,
+            ),
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py]: Addressed in 878ebea and b7f61b8. The full token-tuple Counter was replaced by a constant-size order-independent fingerprint, source tokenization is batched, and the effective record batch now scales with maxSeqLength under a 262144 estimated retained-token budget. Regressions cover duplicates, reordering, mutations, and batch limits; GPU smoke run 31156724042 passed.

File: pkg/finetune/target_unsloth.py [hunk 4]
```diff
@@ -186,6 +205,476 @@ def dataset_load_spec(source: str) -> DatasetLoadSpec:
...
+            first_completion = sequence_completion_mask.index(1)
+            if sequence_completion_mask != [0] * first_completion + [1] * (
+                len(sequence_completion_mask) - first_completion
+            ):
+                raise RuntimeError(
+                    f"prompt-completion preprocessing {location} must mask a prompt prefix and completion suffix"
+                )
+            if remaining_prompt_prefixes is not None:
+                prompt_prefix = tuple(sequence_input_ids[:first_completion])
+                if remaining_prompt_prefixes[prompt_prefix] <= 0:
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py]: Addressed in b7f61b8. The preflight no longer requires standalone prompt token IDs to be an exact prefix. It truncates combined IDs first, computes the prompt boundary as the minimum of standalone prompt length and retained combined length, then fingerprints the retained combined-sequence prefix exactly like pinned Unsloth. Unit and train_model regressions cover a changed boundary token, and GPU smoke run 31156724042 passed.

File: pkg/finetune/target_unsloth.py [hunk 5]
```diff
@@ -186,6 +204,312 @@ def dataset_load_spec(source: str) -> DatasetLoadSpec:
...
+        description="prompt-completion",
+    )
+    if not expected_prompt_ids or not expected_input_ids:
+        raise RuntimeError(
+            "prompt-completion preprocessing verification text must produce tokens"
+        )
+
+    verification_args = copy.copy(trainer.args)
+    verification_args.max_length = len(expected_input_ids)
+    verification_args.dataset_num_proc = 1
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py:3477]: Addressed in b72b383 and b7f61b8. Validation now inspects every real prepared record or BFD-packed segment at the configured maxSeqLength and rejects an all-zero retained completion mask or a missing supervised EOS. The source preflight also mirrors pinned Unsloth truncation before training. Focused regressions and GPU smoke run 31156724042 pass.

File: pkg/finetune/target_unsloth.py [hunk 6]
```diff
@@ -186,6 +205,476 @@ def dataset_load_spec(source: str) -> DatasetLoadSpec:
...
+    )
+    counts: Counter[tuple[Any, ...]] = Counter()
+    for record_index, record in enumerate(dataset):
+        prompt_ids = tokenize_verification_text(
+            processing_class,
+            record["prompt"],
+            add_special_tokens=add_special_tokens,
+            description=f"record {record_index} prompt",
+        )
+        counts[tuple(prompt_ids)] += 1
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py]: Addressed in 878ebea and b7f61b8. Prompt prefixes are represented by a constant-size fingerprint rather than a corpus-sized token Counter. Tokenizer calls are bounded by records and now adapt to maxSeqLength using an estimated retained-token budget. The 65-record batch regression pins the two tokenized sequences per source record; GPU smoke run 31156724042 passed.

File: pkg/finetune/target_unsloth.py [hunk 7]
```diff
@@ -186,6 +216,641 @@ def dataset_load_spec(source: str) -> DatasetLoadSpec:
...
+            ),
+        )
+        prompt_token_rows = token_rows[: len(prompts)]
+        prompt_completion_token_rows = token_rows[len(prompts) :]
+        for batch_index, (prompt_ids, input_ids) in enumerate(
+            zip(prompt_token_rows, prompt_completion_token_rows)
+        ):
+            record_index = batch_start + batch_index
+            input_ids = input_ids[:max_seq_length]
+            prompt_length = min(len(prompt_ids), len(input_ids))
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py:3363]: No code change is needed for the locked runtime. In trl 0.24.0, tokenize_fn only warns when the standalone prompt IDs differ and still builds the mask from len(prompt_ids); it does not back off one token. More importantly, unsloth 2026.8.1 replaces SFTTrainer._prepare_dataset with unsloth_zoo.dataset_utils.sft_prepare_dataset, which truncates pc_ids and then sets n_prompt = min(len(prompt_ids), len(pc_ids)). AIKit imports Unsloth before TRL, so that replacement is active. The b7f61b8 implementation and changed-boundary regressions intentionally mirror this exact path, and GPU run 31156724042 passed.

File: pkg/finetune/target_unsloth.py [hunk 8]
```diff
@@ -186,6 +216,641 @@ def dataset_load_spec(source: str) -> DatasetLoadSpec:
...
+                input_ids[:prompt_length],
+                description=f"record {record_index} prompt",
+            )
+
+    for record in dataset:
+        if add_special_tokens is None:
+            add_special_tokens = prompt_completion_add_special_tokens(
+                processing_class,
+                record["prompt"],
+            )
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py]: No code change is needed because the locked active Unsloth path does not choose this per example. unsloth_zoo 2026.8.1 derives test_text from the first dataset example, computes add_special_tokens once before defining _tokenize_pc, and the mapped closure reuses that flag for every example. AIKit imports Unsloth before TRL and mirrors that same dataset-level decision. Computing it per record here would diverge from the prepared dataset and create the mismatch described by this comment. The BOS/chat-template regression and GPU run 31156724042 pass.

File: pkg/finetune/target_unsloth.py [hunk 9]
```diff
@@ -42,148 +198,3971 @@ class DatasetLoadSpec(NamedTuple):
...
+def tokenize_verification_texts(
+    processing_class: Any,
+    texts: Sequence[str],
+    *,
+    add_special_tokens: bool,
+    description: str,
+) -> list[list[Any]]:
+    tokenized = processing_class(
+        list(texts),
+        add_special_tokens=add_special_tokens,
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py:3180]: Valid finding. Follow-up PR #837, commit `6d85a68`, bounds every prompt-completion verification tokenizer call with explicit right-side truncation and `maxSeqLength`, rejects tokenizers that ignore the limit, and preserves EOS, completion-mask, fingerprint, and source-policy checks. The exact Python suite passes with 167 tests.
